libsignal-service/src/configuration.rs

```diff
@@ -25,22 +25,31 @@ pub struct Credentials {
     pub e164: String,
     pub password: Option<String>,
     pub signaling_key: Option<SignalingKey>,
+    pub device_id: Option<i32>,
 }
 
 impl Credentials {
     /// Kind-of equivalent with `PushServiceSocket::getAuthorizationHeader`
     ///
     /// None when `self.password == None`
-    pub fn authorization(&self) -> Option<(&str, &str)> {
+    pub fn authorization(&self) -> Option<(String, &str)> {
         let identifier = self.login();
         Some((identifier, self.password.as_ref()?))
     }
 
-    pub fn login(&self) -> &str {
-        if let Some(uuid) = self.uuid.as_ref() {
-            uuid
-        } else {
-            &self.e164
+    pub fn login(&self) -> String {
+        let identifier = {
+            if let Some(uuid) = self.uuid.as_ref() {
+                uuid
+            } else {
+                &self.e164
+            }
+            .to_owned()
+        };
+
+        match self.device_id {
+            None | Some(DEFAULT_DEVICE_ID) => identifier,
+            Some(id) => identifier + "." + id.to_string().as_str(),
```

Comment:
```suggestion
            Some(id) => format!("{}.{}", identifier, id),
```

Looks cleaner I think.

Reply:
> I'd still like to know what @gferon does for this though.

Well, you got it exactly right - I do it exactly like that, except I never somehow thought it could be integrated directly in the lib (see: https://github.com/gferon/presage/blob/wip/src/manager.rs#L112)

Thanks a lot for this improvement!

Reply:
Glad I could contribute!